src/ORM/RelatedData/StandardRelatedDataService.php

```diff
@@ -0,0 +1,433 @@
+        }
+    }
+
+    /**
+     * Find the table that contains $componentIDField - this is relevant for subclassed DataObjects
+     * that live in the database as two tables that are joined together
+     *
+     * @param string $class
+     * @param string $componentIDField
+     * @return string
+     */
+    private function findTableNameContainingComponentIDField(string $class, string $componentIDField): string
+    {
+        $tableName = '';
+        $candidateClass = $class;
+        while ($candidateClass) {
+            $dbFields = $this->dataObjectSchema->databaseFields($candidateClass, false);
+            if (array_key_exists($componentIDField, $dbFields)) {
+                $tableName = $this->dataObjectSchema->tableName($candidateClass);
+                break;
+            }
+            $candidateClass = get_parent_class($class);
+        }
+        return $tableName;
+    }
+
+    /**
+     * Loop many_many relationships on the DataObject we're getting usage for
+     * This is for the scenario where there is a missing corresponding
+     * belongs_many_many on the opposite object
+     *
+     * @param int[] $classIDs
+     * @param DataObject $record
+     * @param string[] $throughClasses
+     */
+    private function addRelatedManyManys(
+        array &$classIDs,
+        DataObject $record,
+        array &$throughClasses
+    ): void {
+        $class = get_class($record);
+        foreach ($record->manyMany() as $component => $componentClass) {
+            $componentClass = $this->updateComponentClass($componentClass, $throughClasses);
+            $results = $this->fetchManyManyResults($record, $class, $component, false);
+            $this->addResultsToClassIDs($classIDs, $results, $componentClass);
+        }
+    }
+
+    /**
+     * Query the database to retrieve many-many results
+     *
+     * @param DataObject $record - The DataObject whose usage data is being retrieved, usually a File
+     * @param string $class - example: My\App\SomePageType
+     * @param string $component - example: 'SomeFiles' - My\App\SomePageType::SomeFiles()
+     * @param bool $reverse - true: SomePage::SomeFiles(), false: SomeFile::SomePages()
+     * @return Query|null
+     */
+    private function fetchManyManyResults(
+        DataObject $record,
+        string $class,
+        string $component,
+        bool $reverse
+    ): ?Query {
+        // Example php file: class MyPage ... private static $many_many = [ 'MyFile' => File::class ]
+        $data = $this->dataObjectSchema->manyManyComponent($class, $component);
+        if (!$data || !($data['join'] ?? false)) {
+            return null;
+        }
+        $joinTableName = $this->deriveJoinTableName($data);
+        if (!ClassInfo::hasTable($joinTableName)) {
+            return null;
+        }
+        $usesThroughTable = $data['join'] != $joinTableName;
+
+        $parentField = rtrim($data['parentField'], 'ID') . 'ID';
```

Comment:
That's still not going to work if you have a column called something like `MyColumnIDIDIDIDID'

Reply:
Updated to use preg_replace